File: head-first-design-patterns/observer-pattern/weatherData.py
from subject import Subject

class Weather_Data(Subject):
    
    def __init__(self) -> None:
        self.observers = []
        self.temp = 0.0
        self.humidity = 0.0
        self.pressure = 0.0


    def register_Observer(self, Observer):
        self.observers.append(Observer)


    def remove_Observer(self, Observer):
        self.observers.remove(Observer)


    def notify_Observer(self):
        for o in self.observers:
            o.update(self.temp, self.humidity, self.pressure)


    def set_measurements(self, temp, humidity, pressure):
        self.temp = temp
        self.humidity = humidity
        self.pressure = pressure
        self.notify_Observer()

File: head-first-design-patterns/java/observer/WeatherData.java
import java.util.ArrayList;
import java.util.List;

public class WeatherData implements Subject {
    private List<Observer> observers;
    private float tempreature;
    private float humidity;
    private float pressure;

    public WeatherData() {
        observers = new ArrayList<Observer>();
    }

    public void registerObserver(Observer o) {
        observers.add(o);
    }

    public void removeObserver(Observer o) {
        observers.remove(o);
    }

    public void notifyObservers() {
        for (Observer observer: observers) {
            observer.update(tempreature, humidity, pressure);
        }
    }

    public void measurementsChanged() {
        notifyObservers();
    }

    public void setMeasurements(float tempreature, float humidity, float pressure) {
        this.tempreature = tempreature;
        this.humidity = humidity;
        this.pressure = pressure;
        measurementsChanged();
    }
}

